core/src/main/scala/org/apache/spark/deploy/security/HiveDelegationTokenProvider.scala

```diff
@@ -67,11 +67,11 @@ private[spark] class HiveDelegationTokenProvider
     // Other modes (such as client with or without keytab, or cluster mode with keytab) do not need
     // a delegation token, since there's a valid kerberos TGT for the right user available to the
     // driver, which is the only process that connects to the HMS.
-    val deployMode = sparkConf.get("spark.submit.deployMode", "client")
-    UserGroupInformation.isSecurityEnabled &&
+    val currentToken = UserGroupInformation.getCurrentUser().getCredentials().getToken(tokenAlias)
+    currentToken == null && UserGroupInformation.isSecurityEnabled &&
```

Comment:
No, new DTs are create with empty UGI/Credentials instances, where no previous tokens exist, and then new tokens overwrite the "current user" credentials.

Reply:
sorry I'm not sure I follow.  To put it another way -- whats the situation in which you're trying to *avoid* renewing tokens with this change?

it looks to me like the first time, `delegationTokensRequired` will not find an existing token, so will return true, and then obtainDelegationTokens will get DTs with a new Credentials and then copy them to the current user.  But then on the next call to `delegationTokensRequired`, there will still be the old token sitting in the current user, so `currentToken != null`.  That will remain true forever.

Reply:
The Hive provider is a bit weird. The comment above explains a bit of the issue.

The added code avoids trying to get a delegation token in the case where you have a pre-defined token cache (e.g. what happens when you submit via Oozie). That would generate an ugly exception in my tests, although the app itself would work fine.

Hive tokens are actually never re-created by Spark. In client mode, they're not needed, since the driver has a kerberos tgt. In cluster mode, you either have a keytab and thus a tgt, or you have tokens created by the launcher, in which case you won't create new ones.

Reply:
ok that makes more sense.  I think it would help to expand that comment that hive tokens are never renewed, and may even be provided eg. w/ oozie.